File: DSA/Devansh-Nishchal-Mandloi/day10que2.cpp
#include <iostream>
#include <stack>

using namespace std;

class Queue {
private:
    stack<int> stackEnqueue;
    stack<int> stackDequeue;

    void transferStacks() {
        while (!stackEnqueue.empty()) {
            stackDequeue.push(stackEnqueue.top());
            stackEnqueue.pop();
        }
    }

public:
    void enqueue(int x) {
        stackEnqueue.push(x);
    }

    int dequeue() {
        if (stackDequeue.empty()) {
            if (stackEnqueue.empty()) {
                throw out_of_range("Queue is empty");
            }
            transferStacks();
        }
        int front = stackDequeue.top();
        stackDequeue.pop();
        return front;
    }

    bool isEmpty() const {
        return stackEnqueue.empty() && stackDequeue.empty();
    }
};

int main() {
    Queue q;
    q.enqueue(1);
    q.enqueue(2);

    try {
        cout << "Dequeued: " << q.dequeue() << endl; // Output: 1
        cout << "Dequeued: " << q.dequeue() << endl; // Output: 2

        // Uncomment the following line to test dequeue from an empty queue
        // cout << "Dequeued: " << q.dequeue() << endl;
    } catch (const out_of_range& e) {
        cerr << e.what() << endl;
    }

    return 0;
}

File: DSA/Priyanshi/Day10_que2.java
import java.util.*;

class Day10_que2 {
    private Stack<Integer> stack1;
    private Stack<Integer> stack2;

    public Day10_que2() {
        stack1 = new Stack<>();
        stack2 = new Stack<>();
    }

    public void enqueue(int x) {
        stack1.push(x);
    }

    public int dequeue() {
        if (stack2.isEmpty()) {
            while (!stack1.isEmpty()) {
                stack2.push(stack1.pop());
            }
        }

        if (stack2.isEmpty()) {
            System.out.println("Queue is empty");
            return -1;
        }

        return stack2.pop();
    }

    public int peek() {
        if (stack2.isEmpty()) {
            while (!stack1.isEmpty()) {
                stack2.push(stack1.pop());
            }
        }

        if (stack2.isEmpty()) {
            System.out.println("Queue is empty");
            return -1;
        }

        return stack2.peek();
    }

    public boolean isEmpty() {
        return stack1.isEmpty() && stack2.isEmpty();
    }

    public static void main(String[] args) {
        Day10_que2 queue = new Day10_que2();
        
        queue.enqueue(1);
        queue.enqueue(2);
        System.out.println("Dequeued element: " + queue.dequeue()); // Output: 1
        System.out.println("Dequeued element: " + queue.dequeue()); // Output: 2

        queue.enqueue(3);
        queue.enqueue(4);
        System.out.println("Front element: " + queue.peek()); // Output: 3
        System.out.println("Dequeued element: " + queue.dequeue()); // Output: 3
        System.out.println("Front element: " + queue.peek()); // Output: 4
    }
}
